Two fixl throw exception

```aya
data False
def fixl ¬ (A : Type) => A -> False
def NonEmpty (A : Type) => ¬ ¬ A
```

```
Caused by: java.util.NoSuchElementException
	at kala.collection.mutable.MutableLinkedList.removeLast(MutableLinkedList.java:393)
	at kala.collection.mutable.MutableDeque.dequeue(MutableDeque.java:140)
	at org.aya.util.binop.BinOpParser.makeBinApp(BinOpParser.java:144)
	at org.aya.util.binop.BinOpParser.foldTop(BinOpParser.java:138)
	at org.aya.util.binop.BinOpParser.foldLhsFor(BinOpParser.java:132)
	at org.aya.util.binop.BinOpParser.convertToPrefix(BinOpParser.java:85)
	at org.aya.util.binop.BinOpParser.build(BinOpParser.java:53)
	at org.aya.concrete.desugar.Desugarer.pre(Desugarer.java:78)
	at org.aya.concrete.visitor.EndoExpr.apply(EndoExpr.java:25)
	at org.aya.concrete.visitor.EndoExpr.apply(EndoExpr.java:15)
	at kala.control.Either$Left.map(Either.java:166)
	at org.aya.concrete.visitor.StmtConsumer.accept(StmtConsumer.java:25)
	at org.aya.concrete.stmt.Stmt.desugar(Stmt.java:39)
	at org.aya.concrete.stmt.Stmt.lambda$resolve$0(Stmt.java:35)
	at kala.collection.immutable.ImmutableVectors$Vector1.forEach(ImmutableVectors.java:440)
	at org.aya.concrete.stmt.Stmt.resolve(Stmt.java:35)
	at org.aya.resolve.module.ModuleLoader.resolveModule(ModuleLoader.java:56)
	at org.aya.cli.library.LibraryModuleLoader.load(LibraryModuleLoader.java:77)
	at org.aya.resolve.module.CachedModuleLoader.lambda$load$0(CachedModuleLoader.java:32)
	at kala.collection.mutable.MutableTreeMap.getOrPut(MutableTreeMap.java:483)
	at org.aya.resolve.module.CachedModuleLoader.load(CachedModuleLoader.java:32)
	at org.aya.resolve.module.ModuleLoader.load(ModuleLoader.java:63)
	at org.aya.cli.library.LibraryCompiler$LibrarySccTycker.tyckOne(LibraryCompiler.java:323)
	at org.aya.cli.library.LibraryCompiler$LibrarySccTycker.tyckSCC(LibraryCompiler.java:310)
	at org.aya.util.tyck.OrgaTycker.tyckSCC(OrgaTycker.java:25)
	at kala.function.CheckedConsumer.accept(CheckedConsumer.java:46)
	at kala.collection.immutable.ImmutableVectors$Vector1.forEach(ImmutableVectors.java:440)
	at kala.collection.base.Traversable.forEachChecked(Traversable.java:706)
	at org.aya.cli.library.LibraryCompiler.make(LibraryCompiler.java:212)
	at org.aya.cli.library.LibraryCompiler.make(LibraryCompiler.java:182)
	at org.aya.cli.library.LibraryCompiler.make(LibraryCompiler.java:161)
	at org.aya.cli.utils.AyaCompiler.catching(AyaCompiler.java:30)
	at org.aya.cli.library.LibraryCompiler.start(LibraryCompiler.java:147)
	at org.aya.lsp.server.AyaLanguageServer.loadLibrary(AyaLanguageServer.java:201)
	at kala.collection.internal.view.SeqViews$FlatMapped.lambda$iterator$0(SeqViews.java:1142)
	at kala.collection.base.Iterators$3.next(Iterators.java:777)
	at kala.collection.base.Iterators$ConcatAll.hasNext(Iterators.java:2116)
	at kala.collection.immutable.ImmutableVector.from(ImmutableVector.java:155)
	at kala.collection.immutable.ImmutableVector.from(ImmutableVector.java:140)
	at kala.collection.CollectionLike.toImmutableVector(CollectionLike.java:107)
	at kala.collection.CollectionLike.toImmutableSeq(CollectionLike.java:95)
	at org.aya.lsp.server.AyaLanguageServer.reload(AyaLanguageServer.java:192)
	at org.aya.lsp.server.AyaLanguageServer.load(AyaLanguageServer.java:373)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	... 13 more
```

--- a/base/src/main/java/org/aya/concrete/desugar/BinExprParser.java
+++ b/base/src/main/java/org/aya/concrete/desugar/BinExprParser.java
@@ -54,6 +54,10 @@ public final class BinExprParser extends BinOpParser<AyaBinOpSet, Expr, Expr.Nam
     opSet.reporter.report(new OperatorError.Fixity(currentOp, current, topOp, top, pos));
   }
 
+  @Override protected void reportMissingOperand(String op, SourcePos pos) {
+    opSet.reporter.report(new OperatorError.MissingOperand(pos, op));
+  }
+
   @Override protected @Nullable OpDecl underlyingOpDecl(@NotNull Expr.NamedArg elem) {
     var expr = elem.term();
     while (expr instanceof Expr.Lift lift) expr = lift.expr();

--- a/base/src/main/java/org/aya/concrete/desugar/BinPatternParser.java
+++ b/base/src/main/java/org/aya/concrete/desugar/BinPatternParser.java
@@ -6,12 +6,12 @@ import kala.collection.SeqView;
 import kala.value.MutableValue;
 import org.aya.concrete.Pattern;
 import org.aya.concrete.error.OperatorError;
-import org.aya.util.Arg;
 import org.aya.ref.DefVar;
 import org.aya.ref.LocalVar;
 import org.aya.resolve.ResolveInfo;
 import org.aya.resolve.context.Context;
 import org.aya.tyck.pat.PatternProblem;
+import org.aya.util.Arg;
 import org.aya.util.binop.Assoc;
 import org.aya.util.binop.BinOpParser;
 import org.aya.util.binop.BinOpSet;
@@ -57,6 +57,10 @@ public final class BinPatternParser extends BinOpParser<AyaBinOpSet, Pattern, Ar
 
   @Override protected void reportFixityError(Assoc top, Assoc current, String topOp, String currentOp, SourcePos pos) {
     opSet.reporter.report(new OperatorError.Fixity(currentOp, current, topOp, top, pos));
+  }
+
+  @Override protected void reportMissingOperand(String op, SourcePos pos) {
+    opSet.reporter.report(new OperatorError.MissingOperand(pos, op));
   }
 
   @Override protected @NotNull Pattern createErrorExpr(@NotNull SourcePos sourcePos) {

--- a/base/src/main/java/org/aya/concrete/error/OperatorError.java
+++ b/base/src/main/java/org/aya/concrete/error/OperatorError.java
@@ -97,4 +97,11 @@ public interface OperatorError extends Problem {
       );
     }
   }
+
+  record MissingOperand(@NotNull SourcePos sourcePos, @NotNull String op) implements OperatorError {
+    @Override public @NotNull Doc describe(@NotNull DistillerOptions options) {
+      return Doc.sep(Doc.english("There is no operand for this operator"),
+        Doc.styled(Style.code(), op));
+    }
+  }
 }

--- a/base/src/test/resources/failure/operator/issue677.aya
+++ b/base/src/test/resources/failure/operator/issue677.aya
@@ -1,0 +1,3 @@
+data False
+def fixl ¬ (A : Type) => A -> False
+def NonEmpty (A : Type) => ¬ ¬ A

--- a/base/src/test/resources/failure/operator/issue677.aya.txt
+++ b/base/src/test/resources/failure/operator/issue677.aya.txt
@@ -1,0 +1,11 @@
+In file $FILE:3:27 ->
+
+  1 | data False
+  2 | def fixl ¬ (A : Type) => A -> False
+  3 | def NonEmpty (A : Type) => ¬ ¬ A
+                                 ^^
+
+Error: There is no operand for this operator `¬`
+
+1 error(s), 0 warning(s).
+What are you doing?

--- a/tools/src/main/java/org/aya/util/binop/BinOpParser.java
+++ b/tools/src/main/java/org/aya/util/binop/BinOpParser.java
@@ -72,8 +72,10 @@ public abstract class BinOpParser<
         while (opStack.isNotEmpty()) {
           var top = opStack.peek();
           var cmp = opSet.compare(top._2, currentOp);
-          if (cmp == BinOpSet.PredCmp.Tighter) foldLhsFor(expr);
-          else if (cmp == BinOpSet.PredCmp.Equal) {
+          if (cmp == BinOpSet.PredCmp.Tighter) {
+            if (!foldLhsFor(expr))
+              return createErrorExpr(sourcePos);
+          } else if (cmp == BinOpSet.PredCmp.Equal) {
             // associativity should be specified to both left/right when their share
             // the same precedence. Or a parse error should be reported.
             var topAssoc = top._2.assoc();
@@ -82,8 +84,9 @@ public abstract class BinOpParser<
               reportFixityError(topAssoc, currentAssoc, top._2.name(), currentOp.name(), of(top._1));
               return createErrorExpr(sourcePos);
             }
-            if (topAssoc.leftAssoc()) foldLhsFor(expr);
-            else break;
+            if (topAssoc.leftAssoc()) {
+              if (!foldLhsFor(expr)) return createErrorExpr(sourcePos);
+            } else break;
           } else if (cmp == BinOpSet.PredCmp.Looser) {
             break;
           } else {
@@ -106,6 +109,7 @@ public abstract class BinOpParser<
 
   protected abstract void reportAmbiguousPred(String op1, String op2, SourcePos pos);
   protected abstract void reportFixityError(Assoc top, Assoc current, String topOp, String currentOp, SourcePos pos);
+  protected abstract void reportMissingOperand(String op, SourcePos pos);
   protected abstract @NotNull Expr createErrorExpr(@NotNull SourcePos sourcePos);
 
   private @NotNull Seq<Elm> insertApplication() {
@@ -128,39 +132,46 @@ public abstract class BinOpParser<
     return appliedOperands.getOrPut(elem, MutableSet::of);
   }
 
-  private void foldLhsFor(@NotNull Elm forOp) {
-    foldTop();
-    markAppliedOperand(forOp, AppliedSide.Lhs);
+  private boolean foldLhsFor(@NotNull Elm forOp) {
+    var ok = foldTop();
+    if (ok) markAppliedOperand(forOp, AppliedSide.Lhs);
+    return ok;
   }
 
-  private void foldTop() {
-    var op = opStack.pop();
-    prefixes.append(makeBinApp(op._1));
-  }
-
-  private @NotNull Elm makeBinApp(@NotNull Elm op) {
-    var assoc = toSetElem(op, opSet).assoc();
-    if (assoc.isUnary()) {
-      var operand = prefixes.dequeue();
-      return makeArg(union(operand, op), op.term(), operand, op.explicit());
-    } else if (assoc.isBinary()) {
-      if (prefixes.sizeGreaterThanOrEquals(2)) {
-        var rhs = prefixes.dequeue();
-        var lhs = prefixes.dequeue();
-        return makeBinApp(op, rhs, lhs);
-      } else if (prefixes.sizeEquals(1)) {
-        // BinOP section
-        var sides = getAppliedSides(op);
-        var applied = prefixes.dequeue();
-        var side = sides.isEmpty() ? AppliedSide.Lhs : sides.elementAt(0);
-        // ^ a unary operator is used as binary section, report here or in type checker?
-        return makeSectionApp(union(op, applied), op, elem -> (switch (side) {
-          case Lhs -> makeBinApp(op, elem, applied);
-          case Rhs -> makeBinApp(op, applied, elem);
-        }).term());
+  private boolean foldTop() {
+    var opDef = opStack.pop();
+    if (opDef._2.assoc().isBinary()) {
+      prefixes.append(foldTopBinary(opDef._1));
+    } else { // implies isUnary()
+      var op = opDef._1;
+      if (prefixes.isEmpty()) {
+        // we don't support unary section -- just raise an error.
+        reportMissingOperand(opDef._2.name(), op.term().sourcePos());
+        return false;
       }
+      var operand = prefixes.dequeue();
+      var app = makeArg(union(operand, op), op.term(), operand, op.explicit());
+      prefixes.append(app);
     }
+    return true;
+  }
 
+  private @NotNull Elm foldTopBinary(@NotNull Elm op) {
+    if (prefixes.sizeGreaterThanOrEquals(2)) {
+      var rhs = prefixes.dequeue();
+      var lhs = prefixes.dequeue();
+      return makeBinApp(op, rhs, lhs);
+    } else if (prefixes.sizeEquals(1)) {
+      // BinOP section
+      var sides = getAppliedSides(op);
+      var applied = prefixes.dequeue();
+      var side = sides.isEmpty() ? AppliedSide.Lhs : sides.elementAt(0);
+      // ^ a unary operator is used as binary section, report here or in type checker?
+      return makeSectionApp(union(op, applied), op, elem -> (switch (side) {
+        case Lhs -> makeBinApp(op, elem, applied);
+        case Rhs -> makeBinApp(op, applied, elem);
+      }).term());
+    }
     throw new InternalError("unreachable");
   }
 

